libyul/ASTNodeRegistry.cpp

```diff
@@ -0,0 +1,155 @@
+	yulAssert(_idToLabelMapping[0] == 0);
+	yulAssert(_idToLabelMapping[1] == 1);
+	// using vector<uint8_t> over vector<bool>, as the latter is optimized for space-efficiency
+	std::vector<uint8_t> labelVisited (_labels.size(), false);
+	size_t numLabels = 0;
+	for (auto const& id: _idToLabelMapping)
+	{
+		yulAssert(id < _labels.size());
+		// it is possible to have multiple references to empty / ghost
+		yulAssert(
+			id < 2 || !labelVisited[id],
+			fmt::format("NodeId {} (label \"{}\") is not unique.", id, _labels[id])
+		);
+		labelVisited[id] = true;
+		if (id >= 2)
+			++numLabels;
+	}
+	yulAssert(numLabels + 2 == _labels.size(), "Unused labels present.");
+	m_labels = std::move(_labels);
+	m_idToLabelMapping = std::move(_idToLabelMapping);
+}
+
+size_t ASTNodeRegistry::idToLabelIndex(NodeId const _id) const
+{
+	yulAssert(_id < m_idToLabelMapping.size());
+	return m_idToLabelMapping[_id];
+}
+
+std::string_view ASTNodeRegistry::operator[](NodeId const _id) const
+{
+	auto const labelIndex = idToLabelIndex(_id);
+	if (labelIndex == ghostId())
+		return lookupGhost(_id);
+	return m_labels[labelIndex];
+}
+
+std::optional<ASTNodeRegistry::NodeId> ASTNodeRegistry::findIdForLabel(std::string_view const _label) const {
+	if (_label.empty())
+		return emptyId();
+	if (_label == ghostPlaceholder)
+		return NodeId{1};
+	for (NodeId id = 2; id <= maximumId(); ++id)
+		if ((*this)[id] == _label)
+			return id;
+	return std::nullopt;
+}
+
+std::string_view ASTNodeRegistry::lookupGhost(NodeId const _id) const
+{
+	yulAssert(idToLabelIndex(_id) == ghostId());
+	auto const [it, _] = m_ghostLabelCache.try_emplace(_id, fmt::format("GHOST[{}]", _id));
+	return it->second;
+}
+
+ASTNodeRegistryBuilder::LabelToIdMapping::LabelToIdMapping():
+	m_mapping{{"", 0}, {ASTNodeRegistry::ghostPlaceholder, 1}}
+{}
+
+std::tuple<ASTNodeRegistry::NodeId, bool> ASTNodeRegistryBuilder::LabelToIdMapping::tryInsert(
+	std::string_view const _label,
+	ASTNodeRegistry::NodeId const _id
+)
+{
+	yulAssert(_label != ASTNodeRegistry::ghostPlaceholder);
+	auto const [it, emplaced] = m_mapping.try_emplace(std::string{_label}, _id);
+	return std::make_tuple(it->second, emplaced);
+}
+
+ASTNodeRegistryBuilder::ASTNodeRegistryBuilder():
+	m_nextId(2)
+{}
+
+ASTNodeRegistryBuilder::ASTNodeRegistryBuilder(ASTNodeRegistry const& _existingRegistry)
+{
+	for (size_t i = 2; i <= _existingRegistry.maximumId(); ++i)
+	{
+		auto const existingLabel = _existingRegistry[i];
+		if (!existingLabel.empty())
+		{
+			auto const [_, inserted] = m_mapping.tryInsert(_existingRegistry[i], i);
+			yulAssert(inserted);
+		}
+	}
+	m_nextId = _existingRegistry.maximumId() + 1;
+}
+
+ASTNodeRegistry::NodeId ASTNodeRegistryBuilder::define(std::string_view const _label)
+{
+	auto const [id, inserted] = m_mapping.tryInsert(_label, m_nextId);
+	if (inserted)
+		m_nextId++;
+	return id;
+}
+
+ASTNodeRegistry ASTNodeRegistryBuilder::build() const
+{
+	auto const& mapping = m_mapping.data();
```

Comment:
No type, generic name. The same name is even used twice for different things without providing any distinction. This is a great example of what bothers me the most about `auto`: hiding of all useful details that makes reading the code janky. Instead of having all the info at hand, it forces you to jump to get it. Not having the type here really hurts readability if you don't include enough info in names themselves:
```suggestion
	auto const& labelToIdMapping = m_definedLabels.labelToIdMapping();
```

Reply:
True, that is poor naming on my end.